packages/yarnpkg-pnpify/sources/Hoister.ts

```diff
@@ -0,0 +1,289 @@
+ * representation in-memory.
+ */
+export class Hoister {
+  /** Raw hoister which does all the hoisting heavy lifting */
+  private readonly rawHoister = new RawHoister();
+
+  /** Hoister options */
+  private readonly options: HoisterOptions;
+
+  public constructor(options: HoisterOptions = {optimizeSizeOnDisk: false}) {
+    this.options = options;
+  }
+
+  /**
+   * Traverses PnP tree and produces input for the `RawHoister`
+   *
+   * @param pnp PnP API
+   *
+   * @returns package tree, packages info and locators
+   */
+  private buildPackageTree(pnp: PnpWalkApi): { packageTree: ReadonlyPackageTree, packages: PackageInfo[], locators: PackageLocator[] } {
+    const packageTree: PackageTree = [];
+    const packages: PackageInfo[] = [];
+    const locators: PackageLocator[] = [];
+    const locatorToPackageMap = new Map<LocatorKey, PackageId>();
+    const packageInfos: PackageInformation<NativePath>[] = [];
+
+    const pnpRoots = pnp.getDependencyTreeRoots ? pnp.getDependencyTreeRoots() : [pnp.topLevel];
+
+    let lastPkgId = 0;
+
+    const getLocatorKey = (locator: PackageLocator): LocatorKey => `${locator.name}:${locator.reference}`;
+
+    const assignPackageId = (locator: PackageLocator, pkg: PackageInformation<NativePath>) => {
+      const locatorKey = getLocatorKey(locator);
+      let pkgId = locatorToPackageMap.get(locatorKey);
+      if (!pkgId) {
+        pkgId = lastPkgId;
+        lastPkgId++;
+        const packagePath = npath.toPortablePath(pkg.packageLocation);
+        let weight = 1;
+        if (this.options.optimizeSizeOnDisk && packagePath.indexOf(`.zip${NODE_MODULES_SUFFIX}`) > 0)
+          weight = fs.statSync(packagePath.split(NODE_MODULES_SUFFIX)[0]).size;
+
+        locatorToPackageMap.set(locatorKey, pkgId);
+        locators.push(locator);
+        packages.push({name: locator.name!, weight});
+        packageInfos.push(pkg);
+      }
+      return pkgId;
+    };
+
+    const getPackageId = (locator: PackageLocator) => locatorToPackageMap.get(getLocatorKey(locator));
+
+    const addedIds = new Set<PackageId>();
+    const addPackageToTree = (pkg: PackageInformation<NativePath>, pkgId: PackageId, parentDepIds: Set<PackageId>) => {
+      addedIds.add(pkgId);
+      const deps = new Set<PackageId>();
+      const peerDeps = new Set<PackageId>();
+      packageTree[pkgId] = {deps, peerDeps};
+
+      for (const [name, reference] of pkg.packageDependencies) {
+        if (reference) {
+          const locator = typeof reference === 'string' ? {name, reference} : {name: reference[0], reference: reference[1]};
+          const depPkgId = getPackageId(locator);
+          if (depPkgId) {
+            if (pkg.packagePeers) {
+              if (pkg.packagePeers.has(name)) {
+                peerDeps.add(depPkgId);
+              } else {
+                deps.add(depPkgId);
+              }
+            } else {
+              if (parentDepIds.has(depPkgId) && depPkgId !== pkgId) {
```

Comment:
That's the "legacy support mode", right? We should add a comment to mention that this heuristic produces subpar results and is only there to support v1 installs.

Reply:
I'd also suggest to abstract that in a variable so that we can factor the content:

```ts
const isPeerDependency = pkg.packagePeers
  ? pkg.packagePeers.has(name)
  : parentDepIds.has(depPkgId) && depPkgId !== pkgId;

if (isPeerDependency) {
  peerDeps.add(depPkgId);
} else {
  deps.add(depPkgId);
}
```

And shouldn't it be `depPkgId === pkgId` btw?

Reply:
Removed legacy support mode